homeassistant/components/media_player/roku.py

```diff
@@ -1,117 +1,89 @@
+    CONF_NAME, CONF_HOST, STATE_IDLE, STATE_PLAYING, STATE_UNKNOWN,
+    STATE_HOME)
 
 REQUIREMENTS = ['python-roku==3.1.5']
 
-KNOWN_HOSTS = []
-DEFAULT_PORT = 8060
+DEPENDENCIES = ['roku']
 
-NOTIFICATION_ID = 'roku_notification'
-NOTIFICATION_TITLE = 'Roku Media Player Setup'
+DEFAULT_PORT = 8060
 
 _LOGGER = logging.getLogger(__name__)
 
 SUPPORT_ROKU = SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK |\
     SUPPORT_PLAY_MEDIA | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE |\
     SUPPORT_SELECT_SOURCE | SUPPORT_PLAY
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
-    vol.Optional(CONF_HOST): cv.string,
-})
-
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the Roku platform."""
-    hosts = []
-
-    if discovery_info:
-        host = discovery_info.get("host")
+    if not discovery_info:
+        return
 
-        if host in KNOWN_HOSTS:
-            return
+    # Manage entity cache for service handler
+    if DATA_ENTITIES not in hass.data:
+        hass.data[DATA_ENTITIES] = []
 
-        _LOGGER.debug("Discovered Roku: %s", host)
-        hosts.append(discovery_info.get("host"))
-
-    elif CONF_HOST in config:
-        hosts.append(config.get(CONF_HOST))
-
-    rokus = []
-    for host in hosts:
-        new_roku = RokuDevice(host)
-
-        try:
-            if new_roku.name is not None:
-                rokus.append(RokuDevice(host))
-                KNOWN_HOSTS.append(host)
-            else:
-                _LOGGER.error("Unable to initialize roku at %s", host)
+    name = discovery_info[CONF_NAME]
+    host = discovery_info[CONF_HOST]
+    entity = RokuDevice(host, name)
 
-        except AttributeError:
-            _LOGGER.error("Unable to initialize roku at %s", host)
-            hass.components.persistent_notification.create(
-                'Error: Unable to initialize roku at {}<br />'
-                'Check its network connection or consider '
-                'using auto discovery.<br />'
-                'You will need to restart hass after fixing.'
-                ''.format(config.get(CONF_HOST)),
-                title=NOTIFICATION_TITLE,
-                notification_id=NOTIFICATION_ID)
+    if entity not in hass.data[DATA_ENTITIES]:
+        hass.data[DATA_ENTITIES].append(entity)
 
-    add_devices(rokus)
+    add_devices([entity])
 
 
 class RokuDevice(MediaPlayerDevice):
     """Representation of a Roku device on the network."""
 
-    def __init__(self, host):
+    def __init__(self, host, name):
         """Initialize the Roku device."""
         from roku import Roku
 
-        self.roku = Roku(host)
-        self.ip_address = host
-        self.channels = []
-        self.current_app = None
-        self.device_info = {}
-
+        self._roku = Roku(host)
+        self._ip_address = host
+        self._channels = []
+        self._current_app = None
+        self._device_info = {}
+        self._name = name
         self.update()
```

Comment:
Remove this update and use `add_device([], True)`

Reply:
@pvizeli I don't understand.  Do you have an example of a component doing this?

Reply:
Look into other platforms